lib/net.js

```diff
@@ -1326,84 +1326,70 @@ function listen(self, address, port, addressType, backlog, fd, exclusive) {
 
 
 Server.prototype.listen = function() {
-  var self = this;
+  var args = new Array(arguments.length);
+  for (var i = 0; i < arguments.length; i++)
+    args[i] = arguments[i];
+  args = normalizeConnectArgs(args);
 
-  var lastArg = arguments[arguments.length - 1];
-  if (typeof lastArg === 'function') {
-    self.once('listening', lastArg);
+  if (args[1]) {
+    this.once('listening', args[1]);
   }
 
-  var port = toNumber(arguments[0]);
+  var options = args[0];
+  if (arguments.length === 0 || typeof arguments[0] === 'function') {
+    // Bind to a random port.
+    options.port = 0;
+  }
 
   // The third optional argument is the backlog size.
   // When the ip is omitted it can be the second argument.
   var backlog = toNumber(arguments[1]) || toNumber(arguments[2]);
```

Comment:
We have the data on _two_ places, `args` and `arguments`. This causes confusion on where things are defaulted, etc, making the code hard to read.